File: py/src/imbrace/resources/financial_documents.py
from typing import Any, Awaitable, Callable, Dict, Optional, TypeVar
from ..http import HttpTransport, AsyncHttpTransport


class FinancialDocumentsNotDeployedError(Exception):
    """Raised when a Financial Documents endpoint returns the FastAPI default
    "Not Found" 404 — meaning the route is not registered on the gateway.

    Distinct from "ID not found" (which has a custom error message).
    """

    def __init__(self, endpoint: str):
        self.endpoint = endpoint
        super().__init__(
            f"Financial Documents endpoint \"{endpoint}\" is not deployed on this "
            f"gateway. These methods (get_file/get_report/list_errors/suggest/fix/"
            f"reset/update_*/delete_*) belong to a backend module that may not be "
            f"exposed on every iMBRACE environment. Contact the iMBRACE backend team "
            f"or use a gateway URL that fronts the Financial Management module."
        )


def _is_not_deployed_error(exc: BaseException) -> bool:
    """Detect FastAPI default 'Not Found' 404 wrapped in ApiError."""
    msg = str(exc)
    return "[404]" in msg and '"detail":"Not Found"' in msg


T = TypeVar("T")


def _wrap(endpoint: str, fn: Callable[[], T]) -> T:
    """Sync: run fn, convert FastAPI default 404 → FinancialDocumentsNotDeployedError."""
    try:
        return fn()
    except Exception as e:
        if _is_not_deployed_error(e):
            raise FinancialDocumentsNotDeployedError(endpoint) from e
        raise


async def _awrap(endpoint: str, fn: Callable[[], Awaitable[T]]) -> T:
    """Async: run awaitable fn, convert FastAPI default 404 → FinancialDocumentsNotDeployedError."""
    try:
        return await fn()
    except Exception as e:
        if _is_not_deployed_error(e):
            raise FinancialDocumentsNotDeployedError(endpoint) from e
        raise


class FinancialDocumentsResource:
    """Financial Documents — Sync.

    Multi-step review/correction workflow for documents extracted via
    :meth:`ChatAiResource.process_document`.

    All methods route to ``/v2/ai/financial_documents/*``. **EXPERIMENTAL**:
    requires the Financial Management backend module to be deployed on your
    gateway. If not deployed, calls raise
    :class:`FinancialDocumentsNotDeployedError` with a clear message.

    For one-shot extraction, use :meth:`ChatAiResource.process_document`.
    """

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    @property
    def _v2(self) -> str:
        return f"{self._base}/v2/ai/financial_documents"

    # ── File / Report — experimental ─────────────────────────────────────────

    def get_file(self, id: str, page: Optional[int] = None, limit: Optional[int] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if page is not None: params["page"] = page
        if limit is not None: params["limit"] = limit
        return _wrap(
            f"GET /v2/ai/financial_documents/{id}",
            lambda: self._http.request("GET", f"{self._v2}/{id}", params=params).json(),
        )

    def get_report(self, id: str, page: Optional[int] = None, limit: Optional[int] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if page is not None: params["page"] = page
        if limit is not None: params["limit"] = limit
        return _wrap(
            f"GET /v2/ai/financial_documents/reports/{id}",
            lambda: self._http.request("GET", f"{self._v2}/reports/{id}", params=params).json(),
        )

    # ── Errors — experimental ────────────────────────────────────────────────

    def list_errors(self, file_id: str, limit: int = -1) -> Dict[str, Any]:
        return _wrap(
            f"GET /v2/ai/financial_documents/errors-files/{file_id}",
            lambda: self._http.request("GET", f"{self._v2}/errors-files/{file_id}", params={"limit": limit}).json(),
        )

    # ── Suggest / Fix / Reset — experimental ─────────────────────────────────

    def suggest(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return _wrap(
            "POST /v2/ai/financial_documents/suggest",
            lambda: self._http.request("POST", f"{self._v2}/suggest", json=body).json(),
        )

    def fix(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return _wrap(
            "POST /v2/ai/financial_documents/fix",
            lambda: self._http.request("POST", f"{self._v2}/fix", json=body).json(),
        )

    def reset(self, body: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return _wrap(
            "POST /v2/ai/financial_documents/reset",
            lambda: self._http.request("POST", f"{self._v2}/reset", json=body or {}).json(),
        )

    # ── Update — experimental ────────────────────────────────────────────────

    def update_file(self, id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return _wrap(
            f"PUT /v2/ai/financial_documents/{id}",
            lambda: self._http.request("PUT", f"{self._v2}/{id}", json=body).json(),
        )

    def update_report(self, id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return _wrap(
            f"PUT /v2/ai/financial_documents/reports/{id}",
            lambda: self._http.request("PUT", f"{self._v2}/reports/{id}", json=body).json(),
        )

    # ── Delete — experimental ────────────────────────────────────────────────

    def delete_file(self, id: str) -> None:
        _wrap(
            f"DELETE /v2/ai/financial_documents/{id}",
            lambda: self._http.request("DELETE", f"{self._v2}/{id}"),
        )

    def delete_report(self, id: str) -> None:
        _wrap(
            f"DELETE /v2/ai/financial_documents/reports/{id}",
            lambda: self._http.request("DELETE", f"{self._v2}/reports/{id}"),
        )


class AsyncFinancialDocumentsResource:
    """Financial Documents — Async. See :class:`FinancialDocumentsResource` for full docs."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    @property
    def _v2(self) -> str:
        return f"{self._base}/v2/ai/financial_documents"

    async def get_file(self, id: str, page: Optional[int] = None, limit: Optional[int] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if page is not None: params["page"] = page
        if limit is not None: params["limit"] = limit

        async def _fn() -> Dict[str, Any]:
            r = await self._http.request("GET", f"{self._v2}/{id}", params=params)
            return r.json()
        return await _awrap(f"GET /v2/ai/financial_documents/{id}", _fn)

    async def get_report(self, id: str, page: Optional[int] = None, limit: Optional[int] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if page is not None: params["page"] = page
        if limit is not None: params["limit"] = limit

        async def _fn() -> Dict[str, Any]:
            r = await self._http.request("GET", f"{self._v2}/reports/{id}", params=params)
            return r.json()
        return await _awrap(f"GET /v2/ai/financial_documents/reports/{id}", _fn)

    async def list_errors(self, file_id: str, limit: int = -1) -> Dict[str, Any]:
        async def _fn() -> Dict[str, Any]:
            r = await self._http.request("GET", f"{self._v2}/errors-files/{file_id}", params={"limit": limit})
            return r.json()
        return await _awrap(f"GET /v2/ai/financial_documents/errors-files/{file_id}", _fn)

    async def suggest(self, body: Dict[str, Any]) -> Dict[str, Any]:
        async def _fn() -> Dict[str, Any]:
            r = await self._http.request("POST", f"{self._v2}/suggest", json=body)
            return r.json()
        return await _awrap("POST /v2/ai/financial_documents/suggest", _fn)

    async def fix(self, body: Dict[str, Any]) -> Dict[str, Any]:
        async def _fn() -> Dict[str, Any]:
            r = await self._http.request("POST", f"{self._v2}/fix", json=body)
            return r.json()
        return await _awrap("POST /v2/ai/financial_documents/fix", _fn)

    async def reset(self, body: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        async def _fn() -> Dict[str, Any]:
            r = await self._http.request("POST", f"{self._v2}/reset", json=body or {})
            return r.json()
        return await _awrap("POST /v2/ai/financial_documents/reset", _fn)

    async def update_file(self, id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        async def _fn() -> Dict[str, Any]:
            r = await self._http.request("PUT", f"{self._v2}/{id}", json=body)
            return r.json()
        return await _awrap(f"PUT /v2/ai/financial_documents/{id}", _fn)

    async def update_report(self, id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        async def _fn() -> Dict[str, Any]:
            r = await self._http.request("PUT", f"{self._v2}/reports/{id}", json=body)
            return r.json()
        return await _awrap(f"PUT /v2/ai/financial_documents/reports/{id}", _fn)

    async def delete_file(self, id: str) -> None:
        async def _fn() -> None:
            await self._http.request("DELETE", f"{self._v2}/{id}")
        await _awrap(f"DELETE /v2/ai/financial_documents/{id}", _fn)

    async def delete_report(self, id: str) -> None:
        async def _fn() -> None:
            await self._http.request("DELETE", f"{self._v2}/reports/{id}")
        await _awrap(f"DELETE /v2/ai/financial_documents/reports/{id}", _fn)

File: ts/src/resources/financial-documents.ts
import { HttpTransport } from "../http.js"

// ── Interfaces ────────────────────────────────────────────────────────────────

export interface PaginationMeta {
  total?: string | number
  page?: number
  limit?: number
  total_pages?: number
  [key: string]: unknown
}

export interface FinancialFileDetail {
  message?: string
  importedFile?: { data: Record<string, unknown>[]; pagination: PaginationMeta }
  report?: { data: Record<string, unknown>[]; pagination: PaginationMeta }
  [key: string]: unknown
}

export interface FinancialReportDetail {
  data: Record<string, unknown>[]
  pagination?: PaginationMeta
  [key: string]: unknown
}

export interface FinancialFileError {
  _id?: string
  file_id?: string
  row?: number
  column?: string
  message?: string
  [key: string]: unknown
}

export interface FinancialErrorListResponse {
  data: FinancialFileError[]
  total?: number
  [key: string]: unknown
}

export interface SuggestInput {
  file_id?: string
  errors?: Record<string, unknown>[]
  [key: string]: unknown
}

export interface SuggestResponse {
  suggestions?: Record<string, unknown>[]
  [key: string]: unknown
}

export interface FixInput {
  file_id?: string
  fixes?: Record<string, unknown>[]
  [key: string]: unknown
}

export interface FixResponse {
  success?: boolean
  fixed?: number
  [key: string]: unknown
}

export interface ResetInput {
  file_id?: string
  [key: string]: unknown
}

export interface ResetResponse {
  success?: boolean
  [key: string]: unknown
}

export interface UpdateFileInput { [key: string]: unknown }
export interface UpdateReportInput { [key: string]: unknown }

// ── Error ─────────────────────────────────────────────────────────────────────

/**
 * Thrown when a Financial Documents endpoint returns the FastAPI default
 * "Not Found" 404 — meaning the route is not registered on the gateway.
 *
 * Distinct from "ID not found" (which has a custom error message).
 */
export class FinancialDocumentsNotDeployedError extends Error {
  constructor(public readonly endpoint: string) {
    super(
      `Financial Documents endpoint "${endpoint}" is not deployed on this gateway. ` +
      `These methods (getFile/getReport/listErrors/suggest/fix/reset/update*/delete*) ` +
      `belong to a backend module that may not be exposed on every iMBRACE environment. ` +
      `Contact the iMBRACE backend team or use a gateway URL that fronts the Financial ` +
      `Management module.`
    )
    this.name = "FinancialDocumentsNotDeployedError"
  }
}

async function withNotDeployedCheck<T>(endpoint: string, fn: () => Promise<T>): Promise<T> {
  try {
    return await fn()
  } catch (e: any) {
    const msg = String(e?.message ?? "")
    if (msg.includes("[404]") && msg.includes('"detail":"Not Found"')) {
      throw new FinancialDocumentsNotDeployedError(endpoint)
    }
    throw e
  }
}

// ── Resource ──────────────────────────────────────────────────────────────────

/**
 * Financial Documents — multi-step review/correction workflow for documents
 * extracted via {@link ChatAiResource.processDocument}.
 *
 * All methods route to `/v2/ai/financial_documents/*`.
 *
 * **@experimental**: requires the Financial Management backend module to be
 * deployed on your gateway. If not deployed, calls throw
 * {@link FinancialDocumentsNotDeployedError} with a clear message.
 *
 * For one-shot extraction, use {@link ChatAiResource.processDocument} instead.
 */
export class FinancialDocumentsResource {
  /**
   * @param base - AI service base URL (gateway root).
   *   All methods append `/v2/ai/financial_documents/...`.
   */
  constructor(
    private readonly http: HttpTransport,
    private readonly base: string,
  ) {}

  private get v2() { return `${this.base.replace(/\/$/, "")}/v2/ai/financial_documents` }

  // ── File / Report detail (paginated) — @experimental ───────────────────────

  /** @experimental */
  async getFile(id: string, params?: { page?: number; limit?: number }): Promise<FinancialFileDetail> {
    const url = new URL(`${this.v2}/${id}`)
    if (params?.page  !== undefined) url.searchParams.set("page",  String(params.page))
    if (params?.limit !== undefined) url.searchParams.set("limit", String(params.limit))
    return withNotDeployedCheck(`GET ${url.pathname}`, async () =>
      this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
    )
  }

  /** @experimental */
  async getReport(id: string, params?: { page?: number; limit?: number }): Promise<FinancialReportDetail> {
    const url = new URL(`${this.v2}/reports/${id}`)
    if (params?.page  !== undefined) url.searchParams.set("page",  String(params.page))
    if (params?.limit !== undefined) url.searchParams.set("limit", String(params.limit))
    return withNotDeployedCheck(`GET ${url.pathname}`, async () =>
      this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
    )
  }

  // ── Errors — @experimental ─────────────────────────────────────────────────

  /** @experimental */
  async listErrors(fileId: string, params?: { limit?: number }): Promise<FinancialErrorListResponse> {
    const url = new URL(`${this.v2}/errors-files/${fileId}`)
    url.searchParams.set("limit", String(params?.limit ?? -1))
    return withNotDeployedCheck(`GET ${url.pathname}`, async () =>
      this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
    )
  }

  // ── Suggest / Fix / Reset — @experimental ──────────────────────────────────

  /** @experimental */
  async suggest(body: SuggestInput): Promise<SuggestResponse> {
    return withNotDeployedCheck("POST /v2/ai/financial_documents/suggest", async () =>
      this.http.getFetch()(`${this.v2}/suggest`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      }).then(r => r.json())
    )
  }

  /** @experimental */
  async fix(body: FixInput): Promise<FixResponse> {
    return withNotDeployedCheck("POST /v2/ai/financial_documents/fix", async () =>
      this.http.getFetch()(`${this.v2}/fix`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      }).then(r => r.json())
    )
  }

  /** @experimental */
  async reset(body?: ResetInput): Promise<ResetResponse> {
    return withNotDeployedCheck("POST /v2/ai/financial_documents/reset", async () =>
      this.http.getFetch()(`${this.v2}/reset`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body ?? {}),
      }).then(r => r.json())
    )
  }

  // ── Update — @experimental ─────────────────────────────────────────────────

  /** @experimental */
  async updateFile(id: string, body: UpdateFileInput): Promise<FinancialFileDetail> {
    return withNotDeployedCheck(`PUT /v2/ai/financial_documents/${id}`, async () =>
      this.http.getFetch()(`${this.v2}/${id}`, {
        method: "PUT",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      }).then(r => r.json())
    )
  }

  /** @experimental */
  async updateReport(id: string, body: UpdateReportInput): Promise<FinancialReportDetail> {
    return withNotDeployedCheck(`PUT /v2/ai/financial_documents/reports/${id}`, async () =>
      this.http.getFetch()(`${this.v2}/reports/${id}`, {
        method: "PUT",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      }).then(r => r.json())
    )
  }

  // ── Delete — @experimental ─────────────────────────────────────────────────

  /** @experimental */
  async deleteFile(id: string): Promise<void> {
    await withNotDeployedCheck(`DELETE /v2/ai/financial_documents/${id}`, async () => {
      await this.http.getFetch()(`${this.v2}/${id}`, { method: "DELETE" })
    })
  }

  /** @experimental */
  async deleteReport(id: string): Promise<void> {
    await withNotDeployedCheck(`DELETE /v2/ai/financial_documents/reports/${id}`, async () => {
      await this.http.getFetch()(`${this.v2}/reports/${id}`, { method: "DELETE" })
    })
  }
}
